igvc_navigation/src/field_d_planner/main.cpp

```diff
@@ -0,0 +1,268 @@
+
+//-------------------------- Helper Methods ----------------------------//
+
+/**
+Publish expanded nodes for visualization purposes. This is not a subscriber
+callback.
+
+@param[in] inds the indices of Nodes that have been expanded in the graph search
+@param[in] expanded_cloud the PCL pointcloud which expanded node indices
+        should be stored in
+@param[in] the publishes with which to publish the PCL pointcloud of expanded nodes
+*/
+void publish_expanded_set(const std::vector<std::tuple<int, int>>& inds,
+                          pcl::PointCloud<pcl::PointXYZRGB>& expanded_cloud, ros::Publisher& expanded_pub)
+{
+  expanded_cloud.clear();
+  expanded_cloud.header.frame_id = "odom";
+
+  for (std::tuple<int, int> ind : inds)
+  {
+    pcl::PointXYZRGB p;
+
+    p.x = static_cast<float>(std::get<0>(ind) - x_initial) * map->resolution;
+    p.y = static_cast<float>(std::get<1>(ind) - y_initial) * map->resolution;
+    p.z = -0.05f;
+
+    // set the node color to red if its g-value is inf. Otherwise blue.
+    if (planner.getG(Node(ind)) == std::numeric_limits<float>::infinity())
+    {
+      uint8_t r = 255, g = 0, b = 0;
+      uint32_t rgb = ((uint32_t)r << 16 | (uint32_t)g << 8 | (uint32_t)b);
+      p.rgb = *reinterpret_cast<float*>(&rgb);
+    }
+    else
+    {
+      uint8_t r = 0, g = 125, b = 125;
+      uint32_t rgb = ((uint32_t)r << 16 | (uint32_t)g << 8 | (uint32_t)b);
+      p.rgb = *reinterpret_cast<float*>(&rgb);
+    }
+
+    expanded_cloud.points.push_back(p);
+  }
+
+  expanded_pub.publish(expanded_cloud);
+}
+
+//--------------------------- ROS Callbacks ----------------------------//
+
+/**
+    Set the current map to be used by the D* Lite search problem. The initial
+    map is used to perform the first search through the occupancy grid (equivalent
+    to A*). All maps thereafter are used to update edge costs for the search problem.
+*/
+void map_callback(const igvc_msgs::mapConstPtr& msg)
+{
+  std::lock_guard<std::mutex> planning_lock(planning_mutex);
+  map = msg;  // update current map
+
+  if (initialize_graph)
+  {
+    // initial x and y coordinates of the graph search problem
+    x_initial = static_cast<int>(map->x_initial);
+    y_initial = static_cast<int>(map->y_initial);
+    planner.NodeGrid.initializeGraph(map);
+    initialize_graph = false;
+  }
+}
+
+/**
+    Assigns a valid goal to the graph search problem. Goal index obtained by
+    converting from the /map frame goal coordinate to the graph index.
+
+    @param[in] msg the message received on the "/waypoint" topic
+*/
+void waypoint_callback(const geometry_msgs::PointStampedConstPtr& msg)
+{
+  std::lock_guard<std::mutex> lock(planning_mutex);
+
+  int goal_x, goal_y;
+  goal_x = static_cast<int>(std::round(msg->point.x / planner.NodeGrid.Resolution)) + x_initial;
+  goal_y = static_cast<int>(std::round(msg->point.y / planner.NodeGrid.Resolution)) + y_initial;
+
+  std::tuple<int, int> new_goal = std::make_tuple(goal_x, goal_y);
```

Comment:
didn't you make a struct `Position`?

Reply:
A better data structure to use in this case is `Node` since the goal will always be a node lying along the graph's vertices. Implemented.